### ISSUE TITLE
Make axes selection keyword arguments in torch.diagonal and torch.transpose consistent

### ISSUE THREAD

**Issue by BrunoKM** [pytorch/pytorch#62955]
## 🚀 Feature
Make the keyword argument names for the 1st and 2nd dimension on which `torch.diagonal` and `torch.transpose` operate consistent between the two functions. Namely, let the keyword arguments either both be `0`-indexed:

```python
x = torch.randn(5, 3, 5, 2)
y = torch.tranpose(x, dim0=0, dim1=2)
z = torch.diagonal(x, dim0=0, dim1=2)
```
or `1`-indexed, as is the case in `numpy` (e.g. [np.diagonal](https://numpy.org/doc/stable/reference/generated/numpy.diagonal.html)):
```python
x = torch.randn(5, 3, 5, 2)
y = torch.tranpose(x, dim1=0, dim2=2)
z = torch.diagonal(x, dim1=0, dim2=2)
```
## Motivation

Currently, the two functions `torch.diagonal` and `torch.transpose` have different interfaces. The transpose has a 0-indexed interface: `torch.transpose(input, dim0, dim1)`, whereas the diagonal has a 1-indexed interface `torch.diagonal(input, offset=0, dim1=0, dim2=1)`. It should be self-evident why it would be easier to remember the interfaces if they were the same for the two functions.

## Pitch

The proposal is to change the interface of `torch.diagonal` to be inline with that of `torch.transpose` (and `torch.swapdims`) by renaming the keyword arguments: `dim1 -> dim0` and `dim2 -> dim1`.

**Comment by mruberry** [comment]
Thanks for filing this issue @BrunoKM! They should definitely be consistent, but changing this will be a little tricky because it may have backwards compatibility implications.

**Comment by BrunoKM** [comment]
@mruberry, that makes sense. This is very much a long term feature request :)

I guess a gradual change could be possible by allowing both interfaces for a few versions? I.e. calling `torch.diagonal(…, dim0=a, dim1=b)` and `torch.diagonal(…, dim1=a, dim2=b)` are both allowed and do the same thing, but mixing keyword and positional as in `torch.diagonal(…, a, dim1=b)` would result in an error.

**Comment by mruberry** [comment]
Yep, an approach like that and/or preferring new versions of the operators, like torch.linalg.diagonal and torch.linalg.transpose.
